internal/cachevalue/cache.go

```diff
@@ -61,30 +75,42 @@ func New[I any]() *Cache[I] {
 // Get will return a cached value or fetch a new one.
 // If the Update function returns an error the value is forwarded as is and not cached.
 func (t *Cache[I]) Get() (item I, err error) {
-	item, ok := t.get(t.ttl())
-	if ok {
-		return item, nil
+	v := t.valErr.Load()
+	ttl := t.ttl()
+	vTime := t.lastUpdateMs.Load()
+	tNow := time.Now().UnixMilli()
+	if v != nil && tNow-vTime < ttl.Milliseconds() {
+		if v.e == nil {
+			return v.v, nil
+		}
+		if v.e != nil && t.CacheError || t.ReturnLastGood {
+			return v.v, v.e
+		}
 	}
 
-	item, err = t.Update()
-	if err != nil {
-		if t.Relax {
-			// if update fails, return current
-			// cached value along with error.
-			//
-			// Let the caller decide if they want
-			// to use the returned value based
-			// on error.
-			item, ok = t.get(0)
-			if ok {
-				return item, err
-			}
+	// Fetch new value.
+	if t.NoWait && v != nil && tNow-vTime < ttl.Milliseconds()*2 && (v.e == nil || t.CacheError) {
+		if t.updating.TryLock() {
+			go func() {
```

Comment:
@harshavardhana `TryLock` will only work once per cache. So it cannot build up.

This will block just like the existing if the Update function hangs. This however gives a TTL window where it will keep responding.